src/MISC/fix_eco_force.cpp

```diff
@@ -0,0 +1,642 @@
+            }
+          }
+        }
+      }
+
+      // sum contributions
+      for (k = 0; k < 3; ++k) {
+        for (dim = 0; dim < 3; ++dim) {
+          f[i][dim] -= (nbr[i].real_phi[0][k] * gradient_ii_cos[0][k][dim] + nbr[i].imag_phi[0][k] * gradient_ii_sin[0][k][dim] + gradient_ij_vec[0][k][dim] + reciprocal_vectors[1][k][dim] * gradient_ij_sca[1][k]);
+          f[i][dim] += (nbr[i].real_phi[1][k] * gradient_ii_cos[1][k][dim] + nbr[i].imag_phi[1][k] * gradient_ii_sin[1][k][dim] + gradient_ij_vec[1][k][dim] + reciprocal_vectors[0][k][dim] * gradient_ij_sca[0][k]);
+        }
+      }
+    }
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixECOForce::post_force_respa(int vflag, int ilevel, int iloop) {
+  if (ilevel == ilevel_respa) post_force(vflag);
+}
+
+/* ---------------------------------------------------------------------- */
+
+double FixECOForce::compute_scalar() {
+  double added_energy_total;
+  MPI_Allreduce(&added_energy, &added_energy_total, 1, MPI_DOUBLE, MPI_SUM, world);
+  return added_energy_total;
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixECOForce::pack_forward_comm(int n, int *list, double *buf, int pbc_flag, int *pbc) {
+  int ii, i, jj, j, k, num;
+  tagint id;
+
+  tagint *tag = atom->tag;
+  int nlocal = atom->nlocal;
+  int m = 0;
+
+  for (i = 0; i < n; ++i) {
+    k = list[i];
+    num = nbr[k].n;
+    buf[m++] = num;
+    buf[m++] = nbr[k].duchi;
+
+    if (u_0 != 0.0) {
+      for (ii = 0; ii < 2; ++ii) {
+        for (jj = 0; jj < 3; ++jj) {
+          buf[m++] = nbr[i].real_phi[ii][jj];
+          buf[m++] = nbr[i].imag_phi[ii][jj];
+        }
+      }
+    }
+
+    for (j = 0; j < num; ++j) {
+      buf[m++] = nbr[k].delta[j][0];
+      buf[m++] = nbr[k].delta[j][1];
+      buf[m++] = nbr[k].delta[j][2];
+
+      // id stored in buf needs to be global ID
+      // if k is a local atom, it stores local IDs, so convert to global
+      // if k is a ghost atom (already comm'd), its IDs are already global
+
+      id = nbr[k].id[j];
+      if (k < nlocal) id = tag[id];
+      buf[m++] = id;
+    }
+  }
+
+  return m;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixECOForce::unpack_forward_comm(int n, int first, double *buf) {
+  int ii, i, jj, j, num;
+  int last = first + n;
+  int m = 0;
+
+  for (i = first; i < last; ++i) {
+    nbr[i].n = num = static_cast<int> (buf[m++]);
+    nbr[i].duchi = buf[m++];
+
+    if (u_0 != 0.0) {
+      for (ii = 0; ii < 2; ++ii) {
+        for (jj = 0; jj < 3; ++jj) {
+          nbr[i].real_phi[ii][jj] = buf[m++];
+          nbr[i].imag_phi[ii][jj] = buf[m++];
+        }
+      }
+    }
+
+    for (j = 0; j < num; ++j) {
+      nbr[i].delta[j][0] = buf[m++];
+      nbr[i].delta[j][1] = buf[m++];
+      nbr[i].delta[j][2] = buf[m++];
+      nbr[i].id[j] = static_cast<tagint> (buf[m++]);
```

Comment:
see above. all integers that could be 64-bit have to converted back from integer to double and back using the `ubuf` union.